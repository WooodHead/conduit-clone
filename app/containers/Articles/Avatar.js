import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Image = styled.img`
  border-radius: 50%;
  width: 50px;
  height: 50px;
`;

function Avatar(props) {
  return (
    <Image src={props.image} />
  );
}

Avatar.defaultProps = {
  image: 'https://static.productionready.io/images/smiley-cyrus.jpg',
};

Avatar.propTypes = {
  image: PropTypes.string.isRequired,
};

export default Avatar;
