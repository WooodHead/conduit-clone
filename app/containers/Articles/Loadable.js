/**
 *
 * Asynchronously loads the component for Articles
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
