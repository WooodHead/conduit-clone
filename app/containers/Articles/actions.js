/*
 *
 * Articles actions
 *
 */

import {
  FETCH_ARTICLES,
  ARTICLES_LOADED,
  LOAD_ARTICLES_ERROR,
} from './constants';

export function fetchArticles(page) {
  return {
    type: FETCH_ARTICLES,
    page,
  };
}

export function articlesLoaded(payload) {
  const { articles } = payload;
  return {
    type: ARTICLES_LOADED,
    articles,
  };
}

export function articlesLoadingError() {
  return {
    type: LOAD_ARTICLES_ERROR,
  };
}
