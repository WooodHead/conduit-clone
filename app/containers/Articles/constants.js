/*
 *
 * Articles constants
 *
 */

export const FETCH_ARTICLES = 'app/Articles/FETCH_ARTICLES';
export const ARTICLES_LOADED = 'app/Articles/ARTICLES_LOADED';
export const LOAD_ARTICLES_ERROR = 'app/Articles/LOAD_ARTICLES_ERROR';
