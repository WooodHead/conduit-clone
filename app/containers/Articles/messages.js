/*
 * Articles Messages
 *
 * This contains all the text for the Articles component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  header: {
    id: 'app.containers.Articles.header',
    defaultMessage: 'This is Articles container !',
  },
});
