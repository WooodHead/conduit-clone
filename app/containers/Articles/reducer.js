/*
 *
 * Articles reducer
 *
 */

import { fromJS } from 'immutable';
import {
  ARTICLES_LOADED,
  LOAD_ARTICLES_ERROR,
} from './constants';

const initialState = fromJS({
  fetching: true,
  error: false,
  posts: [],
  pageCount: 0,
});

function articlesReducer(state = initialState, action) {
  switch (action.type) {
    case ARTICLES_LOADED:
      return state
        .set('fetching', false)
        .set('pageCount', action.pageCount)
        .set('posts', action.articles);
    case LOAD_ARTICLES_ERROR:
      return state
        .set('fetching', false)
        .set('error', true);
    default:
      return state;
  }
}

export default articlesReducer;
