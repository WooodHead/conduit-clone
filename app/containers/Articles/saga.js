import { takeLatest, call, put } from 'redux-saga/effects';
import { FETCH_ARTICLES } from 'containers/Articles/constants';
import {
  articlesLoaded,
  articlesLoadingError,
} from 'containers/Articles/actions';

import request from 'utils/request';
import { getOffsetLimit } from 'utils/url';

export function* getArticles(params) {
  const options = getOffsetLimit(params.page);
  const apiEndpoint = `https://conduit.productionready.io/api/articles?limit=${options.limit}&offset=${options.offset}`;

  try {
    const repos = yield call(request, apiEndpoint);
    yield put(articlesLoaded(repos));
  } catch (err) {
    yield put(articlesLoadingError());
  }
}

export default function* articlesData() {
  yield takeLatest(FETCH_ARTICLES, getArticles);
}
