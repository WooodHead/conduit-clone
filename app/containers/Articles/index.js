/**
 *
 * Articles
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { push } from 'react-router-redux';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import ReactPaginate from 'react-paginate';
import { Container, Row, Col, Alert, Card, CardTitle, CardText } from 'reactstrap';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import { BASE_LIMIT, getUrlParams } from 'utils/url';
import Loader from 'components/Loader';
import { fetchArticles } from './actions';
import { makeSelectPosts, makeSelectError, makeSelectFetching, makeSelectPageCount } from './selectors';
import reducer from './reducer';
import saga from './saga';
import Wrapper from './Wrapper';

export class Articles extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor() {
    super();
    this.onPageChange = this.onPageChange.bind(this);
  }
  componentDidMount() {
    const value = getUrlParams(this.props.location.search, 'page');
    this.props.onFetchArticles(value);
  }
  onPageChange(data) {
    const setActivePage = data.selected + 1;
    this.props.onFetchArticles(setActivePage);
    this.props.onPageChange(setActivePage);
  }
  getPageCount() {
    const { pageCount } = this.props;
    return pageCount / BASE_LIMIT;
  }
  getForcePage() {
    const setForcePage = getUrlParams(this.props.location.search, 'page');
    return parseInt(setForcePage, 10) - 1;
  }
  renderPost(post, index) {
    const { title, description, createdAt } = post;
    return (
      <Wrapper key={index}>
        <Card body>
          <CardTitle>{title}</CardTitle>
          <CardText>{description}</CardText>
          <CardText>
            <small className="text-muted">{createdAt}</small>
          </CardText>
        </Card>
      </Wrapper>
    );
  }

  renderPosts() {
    return (
      <Container>
        <Row>
          <Col xs="8">
            {this.props.posts.map((post, index) => this.renderPost(post, index))}
            <ReactPaginate
              disableInitialCallback={false}
              previousLabel="previous"
              nextLabel="next"
              breakLabel={<a className="page-link" href="">...</a>}
              breakClassName="page-item"
              forcePage={this.getForcePage()}
              pageCount={this.getPageCount()}
              marginPagesDisplayed={3}
              pageRangeDisplayed={3}
              onPageChange={this.onPageChange}
              containerClassName="pagination"
              subContainerClassName="pages pagination"
              pageClassName="page-item"
              pageLinkClassName="page-link"
              activeClassName="active"
              previousClassName="page-item"
              previousLinkClassName="page-link"
              nextClassName="page-item"
              nextLinkClassName="page-link"
            />
          </Col>
          <Col xs="6">
          </Col>
        </Row>
      </Container>
    );
  }
  render() {
    const { fetching, error, posts } = this.props;
    let content = <Loader />;
    if (!fetching) {
      if (!error) {
        if (posts.length !== 0) {
          content = this.renderPosts();
        } else {
          content = (
            <Alert color="info">
              This is a warning alert — check it out!
            </Alert>
          );
        }
      } else {
        content = (
          <Alert color="danger">
            This is a warning alert — check it out!
          </Alert>
        );
      }
    }
    return (
      <Container>
        <Helmet>
          <title>Articles</title>
          <meta name="description" content="Description of Articles" />
        </Helmet>
        {content}
      </Container>
    );
  }
}

Articles.defaultProps = {
  posts: [],
  pageCount: 0,
};

Articles.propTypes = {
  onFetchArticles: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  location: PropTypes.shape({
    search: PropTypes.string,
  }),
  error: PropTypes.bool.isRequired,
  posts: PropTypes.array.isRequired,
  fetching: PropTypes.bool.isRequired,
  pageCount: PropTypes.number.isRequired,
};

const mapStateToProps = createStructuredSelector({
  posts: makeSelectPosts(),
  error: makeSelectError(),
  fetching: makeSelectFetching(),
  pageCount: makeSelectPageCount(),
});

export function mapDispatchToProps(dispatch) {
  return {
    onFetchArticles: (page) => {
      dispatch(fetchArticles(page));
    },
    onPageChange: (page) => {
      if (page) {
        dispatch(push(`/?page=${page}`));
      }
    },
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'articles', reducer });
const withSaga = injectSaga({ key: 'articles', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(Articles);
