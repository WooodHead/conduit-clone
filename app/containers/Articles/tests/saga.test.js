/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import { FETCH_ARTICLES } from 'containers/Articles/constants';

import {
  articlesLoadingError,
} from 'containers/Articles/actions';

import articlesData, { getArticles } from '../saga';

const response = [{
  title: 'First Title',
}, {
  title: 'Second Title',
}];

describe('getRepos Saga', () => {
  let getArticlesGenerator;

  beforeEach(() => {
    getArticlesGenerator = getArticles({
      page: 1,
    });

    const selectDescriptor = getArticlesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getArticlesGenerator.next(response).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the articlesLoaded action if it requests the data successfully', () => {
    const putDescriptor = getArticlesGenerator.throw().value;
    expect(putDescriptor).toEqual(put(articlesLoadingError()));
  });
});

describe('articlesDataSaga Saga', () => {
  const articlesDataSaga = articlesData();

  it('should start task to watch for FETCH_ARTICLES action', () => {
    const takeLatestDescriptor = articlesDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(FETCH_ARTICLES, getArticles));
  });
});
