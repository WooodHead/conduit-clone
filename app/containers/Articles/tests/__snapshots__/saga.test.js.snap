// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRepos Saga should dispatch the articlesLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://conduit.productionready.io/api/articles?limit=10&offset=0",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getRepos Saga should dispatch the articlesLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "articles": undefined,
      "pageCount": undefined,
      "type": "app/Articles/ARTICLES_LOADED",
    },
    "channel": null,
  },
}
`;
